### PR TITLE
[MRG] Plot prob atlas colorbar

### PR DESCRIPTION
**Pull request by arthurmensch**
Issue #1018. I put the colorbar by default for consistency with `plot_stat_map`

Screenshot
screenshot from 2016-02-26 17-29-29

### CODE REVIEW

File: nilearn/plotting/img_plotting.py [hunk 1]
```diff
@@ -800,13 +801,31 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
         thr = max(thr, 1e-6)
         if view_type == 'continuous':
             display.add_overlay(map_img, threshold=thr,
-                                cmap=cm.alpha_cmap(color))
+                                cmap=cm.alpha_cmap(color),)
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Is this change necessary?

File: nilearn/plotting/img_plotting.py [hunk 2]
```diff
@@ -800,13 +801,31 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
...
+        display._cbar.update_ticks()
+        display._cbar.set_ticks(np.round(np.linspace(1, n_maps,
+                                                     min(n_maps, 10))) + .5)
+        display._cbar.set_ticklabels(np.linspace(1, n_maps,
+                                                 min(n_maps, 10)).astype('int'))
+        left, bottom, width, height = display._colorbar_ax.get_position().bounds
+        display._colorbar_ax.set_position([left, bottom, width, height * 0.95])
+        display._colorbar_ax.annotate('Map #', xy=(1, 1.03), ha='right', va='bottom',
+                                      xycoords='axes fraction')
+        # display._colorbar_ax.set_xlabel('Map #')
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Could you remove the comment line

File: nilearn/plotting/img_plotting.py [hunk 3]
```diff
@@ -815,6 +834,32 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
...
+def _cmap_discretize(cmap, N):
+    """Return a discrete colormap from the continuous colormap cmap.
+
+        cmap: colormap instance, eg. cm.jet.
+        N: number of colors.
+
+    Example
+        x = resize(arange(100), (5,100))
+        djet = cmap_discretize(cm.jet, 5)
+        imshow(x, cmap=djet)
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Could you format this according to the numpy doc standard: a Parameters section, and Example as a section.

The reason that I ask this, is that the docstring is invalid currently.

Also, the examples should be doctests.

File: nilearn/plotting/img_plotting.py [hunk 4]
```diff
@@ -815,6 +834,32 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
...
+    """
+
+    if type(cmap) == str:
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: isinstance(cmap, basestring)

File: nilearn/plotting/img_plotting.py (hunk 3 above)
Comment (arthurmensch) [inline nilearn/plotting/img_plotting.py]: This is unused anyway
Le 26 févr. 2016 17:33, "Gael Varoquaux" <notifications@github.com> a
écrit :

**Comment by GaelVaroquaux** [comment]
Can you also add this functionality to an example? Probably the one that demos plot_prob_atlas.

File: nilearn/plotting/img_plotting.py [hunk 5]
```diff
@@ -649,6 +650,7 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
                     output_file=None, display_mode='ortho',
                     figure=None, axes=None, title=None, annotate=True,
                     draw_cross=True, black_bg='auto', dim=False,
+                    colorbar=True,
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: I think that colorbar should be False by default (and in the example just add one being "True").

Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: The reason that I think that is that it isn't that informative.

File: nilearn/plotting/img_plotting.py [hunk 6]
```diff
@@ -805,8 +806,25 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
             display.add_contours(map_img, levels=[thr],
                                  linewidths=linewidths,
                                  colors=[color], filled=filled,
-                                 alpha=alpha, linestyles='solid')
-
+                                 alpha=alpha, linestyles='solid',)
+    if colorbar:
+        our_cmap = cmap
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: I think that the variable 'our_cmap' is not used.

File: examples/01_plotting/plot_prob_atlas.py [hunk 7]
```diff
@@ -49,6 +51,7 @@
                'ICBM tissues': (icbm['wm'], icbm['gm'], icbm['csf'])}
 
 for name, atlas in sorted(atlas_types.items()):
-    plotting.plot_prob_atlas(atlas, title=name)
+    colorbar = name in ['Harvard_Oxford', 'MSDL', 'Smith2009 10 Brainmap']
+    plotting.plot_prob_atlas(atlas, title=name, colorbar=colorbar)
```
Comment (bthirion) [inline examples/01_plotting/plot_prob_atlas.py]: on my box, this creates grayscale colorbars :-(

File: nilearn/plotting/img_plotting.py [hunk 8]
```diff
@@ -805,8 +806,25 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
...
+        tick_locator = matplotlib.ticker.MaxNLocator(nbins=10)
+        display.locator = tick_locator
+        display._cbar.update_ticks()
+        display._cbar.set_ticks(np.round(np.linspace(1, n_maps,
+                                                     min(n_maps, 10))) + .5)
+        display._cbar.set_ticklabels(np.round(np.linspace(1, n_maps,
+                                                 min(n_maps, 10))).astype('int'))
+        left, bottom, width, height = display._colorbar_ax.get_position().bounds
+        display._colorbar_ax.set_position([left, bottom, width, height * 0.95])
+        display._colorbar_ax.annotate('Map #', xy=(1, 1.03), ha='right', va='bottom',
```
Comment (bthirion) [inline nilearn/plotting/img_plotting.py]: Would it be possible to support user-supplied labels (sorry if my question is silly, I don't know this code well) ?

Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: No because they are too big and we cannot do a good layout.

**Comment by GaelVaroquaux** [comment]
By the way: look at travis: your PR doesn't seem to work on all versions of matplotlib, if I read things right.

File: examples/01_plotting/plot_prob_atlas.py (hunk 7 above)
Comment (GaelVaroquaux) [inline examples/01_plotting/plot_prob_atlas.py]: > on my box, this creates grayscale colorbars :-(

Ouch, we have a problem.

**Comment by dohmatob** [comment]
This already looks great. But ideally, it'd be nice to have the possibility to display user-supplied names (not numbers). Thus the idea of the separate figure to contain the color legends.  Or maybe you could have a horizontal colorbar with tilted texts for the legends ?

**Comment by GaelVaroquaux** [comment]
> But ideally, it'd be nice to have the possibility to display
> user-supplied names (not numbers).

This has already been asked by @bthirion a bit earlier on this PR.

> Thus the idea of the separate figure to contain the color legends.

Indeed, I agree that it would be nice, but it's much more work. And the
minute we do that the users will start wanting it to be on the same
figure.

> Or maybe you could have a horizontal colorbar with tilted texts for the
> legends ?

That doesn't seem practical: we would have to rework completely the
layout code.

In the interest of short cycles, and keeping the code base simple, I
suggest that we stick with the simple solution that we have here. There
still is work to make it work on all the MPL versions that we support
(and @bthirion seems to have a bug with the color). With this PR, it's
easy to create a table to put next to the figure in a paper.

File: examples/01_plotting/plot_prob_atlas.py [hunk 9]
```diff
@@ -49,6 +51,7 @@
                'ICBM tissues': (icbm['wm'], icbm['gm'], icbm['csf'])}
 
 for name, atlas in sorted(atlas_types.items()):
-    plotting.plot_prob_atlas(atlas, title=name)
+    colorbar = name in ['Harvard_Oxford', 'MSDL', 'Smith2009 10 Brainmap']
```
Comment (banilo) [inline examples/01_plotting/plot_prob_atlas.py]: Why not also craddock's altas?

File: nilearn/plotting/img_plotting.py [hunk 10]
```diff
@@ -649,6 +650,7 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
                     output_file=None, display_mode='ortho',
                     figure=None, axes=None, title=None, annotate=True,
                     draw_cross=True, black_bg='auto', dim=False,
+                    colorbar=True,
                     cmap=plt.cm.gist_rainbow, vmin=None, vmax=None,
```
Comment (banilo) [inline nilearn/plotting/img_plotting.py]: Add an optional argument to control ``cmap`` here?

File: nilearn/plotting/img_plotting.py [hunk 11]
```diff
@@ -814,7 +832,6 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
 
     return display
 
-
```
Comment (banilo) [inline nilearn/plotting/img_plotting.py]: merge conflicts?

File: examples/01_plotting/plot_prob_atlas.py (hunk 9 above)
Comment (arthurmensch) [inline examples/01_plotting/plot_prob_atlas.py]: Not the point of this PR

File: nilearn/plotting/img_plotting.py (hunk 10 above)
Comment (arthurmensch) [inline nilearn/plotting/img_plotting.py]: Already there

**Comment by AlexandreAbraham** [comment]
> Indeed, I agree that it would be nice, but it's much more work. And the minute we do that the users will start wanting it to be on the same figure.

It's not that much work, but I agree on the second point. If advanced people really wants it, maybe it's a good candidate for `nilearn_sandbox`?

**Comment by arthurmensch** [comment]
@bthirion can you provide a `pip freeze` to reproduce grayscale colorbar ?

**Comment by arthurmensch** [comment]
Travis is green, comment addressed

**Comment by GaelVaroquaux** [comment]
Overall this is great. I have one nitpick: could we have it so that, instead of having:
"colorbar = name in ['Harvard_Oxford', 'MSDL', 'Smith2009 10 Brainmap']"
we don't display the colorbar in the for loop add a line at the beginning (or the end), in a separate notebook-like cell with explanatory text, that turns explicitly on the colorbar for one of those examples (how about "bm10"?).

The reason is that the statement with the "in", is one extra thing to understand and learn. End users might be real beginners in Python. We want to keep it as simple as possible.

File: nilearn/plotting/img_plotting.py [hunk 12]
```diff
@@ -805,8 +809,28 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
...
+                                 alpha=alpha, linestyles='solid', )
+    if colorbar:
+        display._colorbar = True
+        display._show_colorbar(matplotlib.colors.LinearSegmentedColormap.
+                               from_list('segmented colors',
+                                         color_list, n_maps + 1),
+                               matplotlib.colors.Normalize(1, n_maps + 1))
+        tick_locator = matplotlib.ticker.MaxNLocator(nbins=10)
+        display.locator = tick_locator
+        display._cbar.update_ticks()
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: The code below would be more readable if you created an intermediate variable containing 
```
np.round(np.linspace(1, n_maps, min(n_maps, 10)))
```

**Comment by dohmatob** [comment]
LGTM (except for the "in" comment raised above by @GaelVaroquaux). Great job!

**Comment by bthirion** [comment]
you cant get my config here:
[pip_freeze.txt](https://github.com/nilearn/nilearn/files/151380/pip_freeze.txt)
Thx for taking care of this.

File: nilearn/plotting/img_plotting.py [hunk 13]
```diff
@@ -732,6 +735,8 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
             pronounced effect.
         cmap: matplotlib colormap, optional
             The colormap for the atlas maps
+        colorbar : boolean, optional
+            If True, display a colorbar on the right of the plots.
```
Comment (KamalakerDadi) [inline nilearn/plotting/img_plotting.py]: Could you add a line saying about the representation of colorbar ?

File: nilearn/plotting/tests/test_img_plotting.py [hunk 14]
```diff
@@ -501,6 +501,8 @@ def test_plot_prob_atlas():
                     threshold=0.2)
     # Testing the 4D plot prob atlas with contours
     plot_prob_atlas(img, view_type='continuous')
+    # Testing the 4D plot prob atlas without colormap
```
Comment (KamalakerDadi) [inline nilearn/plotting/tests/test_img_plotting.py]: " Testing the 4D plot prob atlas with colorbar " you mean ?

File: nilearn/plotting/img_plotting.py (hunk 12 above)
Comment (KamalakerDadi) [inline nilearn/plotting/img_plotting.py]: yes

**Comment by KamalakerDadi** [comment]
Could you please also add about this feature in ```whats_new``` doc.

**Comment by KamalakerDadi** [comment]
@arthurmensch  I am sorry for conflicts. I realized lately.

**Comment by GaelVaroquaux** [comment]
@bthirion, do you still see the colorbar in black and white?

**Comment by bthirion** [comment]
Yes, it looks like the colorbar of the anatomical image is plotted.
figure_4

**Comment by arthurmensch** [comment]
@bthirion are you using Python 2.7 on Ubuntu 14.04 ?

**Comment by bthirion** [comment]
Yes

**Comment by arthurmensch** [comment]
I cannot reproduce it on a 14.04 conda configuration. Can you dump a pip freeze when you have time ?

**Comment by bthirion** [comment]
Here it is. Thx for taking care of this.
Best,

Bertrand

Brlapi==0.6.1
Cython==0.23.4
GnuPGInterface==0.3.2
Jinja2==2.7.3
Mako==0.9.1
Markdown==2.4
MarkupSafe==0.23
MySQL-python==1.2.3
PAM==0.4.2
Pillow==2.3.0
PsychoPy==1.83.01
PyAudio==0.2.7
PyOpenGL==3.0.2
PyWavelets==0.2.2
PyYAML==3.10
Pygments==2.0.2
SecretStorage==2.0.0
Shapely==1.3.0
Sphinx==1.2.3
Twisted-Core==13.2.0
Twisted-Names==13.2.0
Twisted-Web==13.2.0
UniConvertor==1.1.4
Unidecode==0.04.12
VTK==5.8.0
adium-theme-ubuntu==0.3.4
apptools==4.1.0
apt-xapian-index==0.45
argparse==1.2.1
astroid==1.0.1
backports.ssl-match-hostname==3.4.0.2
beautifulsoup4==4.2.1
blinker==1.2
certifi==14.05.14
cfflib==2.0.5
chardet==2.0.1
colorama==0.2.5
command-not-found==0.3
configglue==1.1.2
configobj==4.7.2
coverage==3.7.1
cvxopt==1.1.4
cycler==0.10.0
datamerger==1.04
debtagshw==0.1
decorator==3.4.0
defer==1.0.6
dipy==0.8.0
dirspec==13.10
docutils==0.12
duplicity==0.6.23
envisage==4.1.0
epydoc==3.0.1
expyriment==0.8.0
feedgenerator==1.5
fuse-python==0.2.1
future==0.14.3
h5py==2.2.1
html5lib==0.999
httplib2==0.8
ipdbplugin==1.4.1
ipython==2.3.1
joblib==0.9.4
jsonschema==2.4.0
keyring==3.5
launchpadlib==1.10.2
lazr.restfulclient==0.13.3
lazr.uri==1.0.3
line-profiler==1.0
lockfile==0.8
logilab-common==0.61.0
louis==2.5.3
lxml==3.3.3
matplotlib==1.5.1
mayavi==4.1.0
memory-profiler==0.31
mercurial==2.8.2
mistune==0.5.1
modl==0.1
mpltools==0.2.0
networkx==1.8.1
nibabel==2.0.2
nilearn==0.2.4
nipy==0.3.0.dev
nipy-data==0.2
nipy-templates==0.2
nipype==0.10.0
nistats==0.1.0.dev0
nose==1.3.1
numexpr==2.2.2
numpy==1.10.4
oauth==1.0.1
oauthlib==0.6.1
oneconf==0.3.7
openpyxl==1.7.0
opensesame==3.0.7b1
pandas==0.15.2
path.py==5.1
patsy==0.3.0
pdfrw==0.1
pelican==3.3
pep8==1.4.6
pexpect==3.1
piston-mini-client==0.7.5
pp==1.6.4
pprocess==0.5
protobuf==2.5.0
psutil==1.2.1
ptyprocess==0.4
py==1.4.31
pyHANSO==0.1-git
pyOpenSSL==0.13
pycortex==0.1.1
pycrypto==2.6.1
pycups==1.9.66
pycurl==7.19.3
pydicom==0.9.8
pyface==4.1.0
pyflakes==0.8.1
pygame==1.9.1release
pyglet==1.1.4
pygobject==3.12.0
pygpgme==0.3
pyhrf==0.3
pyinotify==0.9.4
pylint==1.1.0
pymvpa2==2.3.1
pyo==0.6.8
pyparallel==0.2
pyparsing==2.0.1
pypreprocess==0.1-git
pyserial==2.6
pysmbc==1.0.14.1
pysurfer==0.5
pytest==2.8.7
python-apt==0.9.3.5ubuntu2
python-dateutil==2.3
python-debian==0.1.21-nmu2ubuntu2
python-qprogedit==3.2.2
pytz==2014.10
pyxdg==0.25
pyzmq==14.4.1
reportlab==3.0
requests==2.2.1
roman==2.0.0
rst2pdf==0.93.dev
rubber==1.1
scikit-image==0.11.3
scikit-learn==0.17.1
scipy==0.17.0
sessioninstaller==0.0.0
simplegeneric==0.8.1
simplejson==3.3.1
six==1.8.0
software-center-aptd-plugins==0.0.0
statsmodels==0.6.1
sympy==0.7.4.1
system-service==0.1.6
tables==3.1.1
terminado==0.5
tornado==4.1
traits==4.1.0
traitsui==4.1.0
tv-l1-solver==0.1-git
unity-lens-photos==1.0
urllib3==1.7.1
virtkey==0.63.0
visionegg==1.2.1
wadllib==1.3.2
webcolors==1.4
wheel==0.24.0
wsgiref==0.1.2
wxPython==2.8.12.1
wxPython-common==2.8.12.1
xdiagnose==3.6.3build2
xkit==0.0.0
xlrd==0.9.2
xlwt==0.7.5
zope.interface==4.0.5

File: nilearn/plotting/img_plotting.py [hunk 15]
```diff
@@ -817,8 +821,28 @@ def plot_prob_atlas(maps_img, anat_img=MNI152TEMPLATE, view_type='auto',
             display.add_contours(map_img, levels=[thr],
                                  linewidths=linewidths,
                                  colors=[color], filled=filled,
-                                 alpha=alpha, linestyles='solid')
-
+                                 alpha=alpha, linestyles='solid', )
```
Comment (KamalakerDadi) [inline nilearn/plotting/img_plotting.py]: Is this change intentional ?

File: examples/01_plotting/plot_prob_atlas.py [hunk 16]
```diff
@@ -22,7 +24,7 @@
 # Load 4D probabilistic atlases
 from nilearn import datasets
 
-# Harvard Oxford Atlas
+# Harvard Oxford Atlasf
```
Comment (KamalakerDadi) [inline examples/01_plotting/plot_prob_atlas.py]: typo here

**Comment by KamalakerDadi** [comment]
@bthirion Can you try with ```view_type='continuous'```. Just want to make sure that if there is any problem with contours method.

```python
plotting.plot_prob_atlas(msdl.maps, colorbar=True, view_type='continuous')
```

**Comment by bthirion** [comment]
This works well with `view_type=continuous`, thx.

**Comment by KamalakerDadi** [comment]
I am not sure but assuming that there might be problem with matplotlib version 1.5.1 using contours. Not guaranteed.

**Comment by arthurmensch** [comment]
@bthirion what matplotlib backend do you use ?

**Comment by bthirion** [comment]
Qt4Agg

**Comment by arthurmensch** [comment]
I cannot reproduce this issue. As it is not critical, I suggest we merge this PR and wait for further feedback to identity the reason for this corner case.
